feat(databases): test connection api

lilykuang opened the pull request:
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
- Add test connection functionality to `/api/v1/database/test_connection`

### TEST PLAN
- python test

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
test connection api on databases

Commit:
update test connection tests

Commit:
update database api test and open api description

willbarrett commented on superset/databases/dao.py:
```diff
@@ -0,0 +1,75 @@
+logger = logging.getLogger(__name__)
+
+
+class DatabaseDAO(BaseDAO):
```
This probably makes more sense as a Command rather than a DAO since it is more of a business operation.

willbarrett commented on superset/databases/dao.py:
```diff
@@ -0,0 +1,75 @@
+        # with the safe URI, we assume we should retrieve the decrypted URI to test
+        # the connection.
+        if db_name:
+            existing_database = (
```
This database lookup would be a good candidate for a DAO method.

willbarrett commented on superset/databases/dao.py:
```diff
@@ -0,0 +1,75 @@
+                uri = existing_database.sqlalchemy_uri_decrypted
+
+        # this is the database instance that will be tested
+        database = Database(
```
Here through line 70 could also potentially be a DAO method - something like `build_db_for_connection_test`

willbarrett commented on superset/databases/api.py:
```diff
@@ -335,3 +351,88 @@ def select_star(
+                encrypted_extra=json.dumps(request.json.get("encrypted_extra", {})),
+            )
+            return self.response(200, message="OK")
+        except CertificateException as ex:
```
This breaks encapsulation a little bit. We're relying on SQLAlchemy errors to decide the response format. Ideally we'd capture these errors in the DAO or Command and transform them into a Superset-specific exception. That could allow us to group, for instance, the module loading errors at a lower level.

willbarrett commented on superset/databases/api.py:
```diff
@@ -335,3 +351,88 @@ def select_star(
+            return self.response(200, message="OK")
+        except CertificateException as ex:
+            logger.info("Certificate exception")
+            return self.response(500, message=ex.message)
```
Since this is a connection test endpoint, let's return a 400 for this instead of a 500. We expect connection failures.

willbarrett commented on superset/views/core.py:
```diff
@@ -1162,7 +1133,7 @@ def testconn(  # pylint: disable=too-many-return-statements,no-self-use
             return json_error_response(_(str(ex)), 400)
         except Exception as ex:  # pylint: disable=broad-except
-            logger.error("Unexpected error %s", type(ex).__name__)
+            logger.warning("Unexpected error %s", type(ex).__name__)
```
👍

willbarrett commented on superset/databases/api.py:
```diff
@@ -335,3 +351,88 @@ def select_star(
+        except OperationalError:
+            logger.warning("Connection failed")
+            return self.response(
+                500, message="Connection failed, please check your connection settings"
```
I think we're missing the `_()` function that will hook these messages up with their associated translations.

willbarrett commented on superset/databases/api.py:
```diff
@@ -335,3 +351,88 @@ def select_star(
+        except (NoSuchModuleError, ModuleNotFoundError):
+            logger.info("Invalid driver")
+            driver_name = make_url(uri).drivername
+            message = "Could not load database driver: %s" % (driver_name)
```
I think we could use an f-string here.

willbarrett reviewed (commented):
A good bit of progress here! I've made a couple of recommendations on code structure. I think we could do a better job of centralizing the error handling coming out of SQLAlchemy to one location. We could potentially raise custom error classes that bring with them an associated response code and message - it looks like the language is the same between the endpoints, and that would DRY things up further. Happy to discuss the feedback if that would be helpful!

Commit:
moved test connection to commands

Commit:
update error message

willbarrett reviewed (approved):
LGTM - thanks for all the work on this!

Commit:
rebase master to test connection branch
# Conflicts:
#	superset/databases/api.py
#	superset/databases/dao.py
#	tests/databases/api_tests.py

Commit:
fix isort

Commit:
fix mypy

Commit:
fix black

Commit:
fix mypy pre commit

riahk commented:
LGTM!!
